cdap-common/src/main/java/io/cdap/cdap/common/schedule/AbstractScheduleClient.java

```diff
@@ -0,0 +1,139 @@
+    .create();
+
+  private static final Type LIST_SCHEDULE_DETAIL_TYPE = new TypeToken<List<ScheduleDetail>>() {
+  }.getType();
+
+  /**
+   * Executes an HTTP request.
+   */
+  protected abstract HttpResponse execute(HttpRequest request, int... allowedErrorCodes)
+    throws IOException, UnauthenticatedException, UnauthorizedException;
+
+  /**
+   * Resolved the specified URL
+   */
+  protected abstract URL resolve(String resource) throws IOException;
+
+  /**
+   * List all schedules for an existing workflow.
+   *
+   * @param workflow the ID of the workflow
+   */
+  public List<ScheduleDetail> listSchedules(WorkflowId workflow)
+    throws IOException, UnauthenticatedException, NotFoundException, UnauthorizedException, BadRequestException {
+    return doList(workflow);
+  }
+
+  private List<ScheduleDetail> doList(WorkflowId workflow)
+    throws IOException, UnauthenticatedException, NotFoundException, UnauthorizedException, BadRequestException {
+    String path = String.format("namespaces/%s/apps/%s/workflows/%s/schedules",
+                                workflow.getNamespace(), workflow.getApplication(), workflow.getProgram());
+    HttpResponse response = makeRequest(path, HttpMethod.GET, null);
+
+    if (HttpURLConnection.HTTP_NOT_FOUND == response.getResponseCode()) {
+      throw new NotFoundException(workflow);
+    }
+
+    ObjectResponse<List<ScheduleDetail>> objectResponse =
+      ObjectResponse.fromJsonBody(response, LIST_SCHEDULE_DETAIL_TYPE, GSON);
+
+    return objectResponse.getResponseObject();
+  }
+
+  /**
+   * Get the next scheduled run time of the program. A program may contain multiple schedules. This method returns the
+   * next scheduled runtimes for all the schedules. This method only takes + into account schedules based on time.
+   * Schedules based on data are ignored.
+   *
+   * @param workflow Id of the Workflow for which to fetch next run times.
+   * @return list of Scheduled runtimes for the Workflow. Empty list if there are no schedules.
+   */
+  public List<ScheduledRuntime> nextRuntimes(WorkflowId workflow)
+    throws IOException, UnauthenticatedException, NotFoundException, UnauthorizedException, BadRequestException {
+
+    String path = String.format("namespaces/%s/apps/%s/workflows/%s/nextruntime",
+                                workflow.getNamespace(), workflow.getApplication(), workflow.getProgram());
+    HttpResponse response = makeRequest(path, HttpMethod.GET, null);
+
+    if (HttpURLConnection.HTTP_NOT_FOUND == response.getResponseCode()) {
+      throw new NotFoundException(workflow);
+    }
+
+    ObjectResponse<List<ScheduledRuntime>> objectResponse = ObjectResponse.fromJsonBody(
+      response, new TypeToken<List<ScheduledRuntime>>() {
+      }.getType(), GSON);
+
+    return objectResponse.getResponseObject();
+  }
+
+  private HttpResponse makeRequest(String path, HttpMethod httpMethod, @Nullable String body)
```

Comment:
the makeRequest methods also seem duplicate

Reply:
I also moved this method to the super class